File: sdk/python/x402/protocol.py
"""x402 protocol header encoding/decoding.

This module uses the native Rust implementation via FFI when available,
falling back to pure Python for compatibility.
"""

import base64
import json
from typing import Any, Tuple

from x402.types import PaymentRequirements, SignedPayment, PaymentPayload

# Try to import native Rust bindings
try:
    from x402_native import (
        encode_requirements as _native_encode_requirements,
        decode_requirements as _native_decode_requirements,
        encode_payment as _native_encode_payment,
        decode_payment as _native_decode_payment,
        PaymentRequirements as NativePaymentRequirements,
        PaymentPayload as NativePaymentPayload,
    )
    _USE_NATIVE = True
except ImportError:
    _USE_NATIVE = False


# Header names
X402_REQUIREMENTS_HEADER = "X-Payment-Requirements"
X402_PAYMENT_HEADER = "X-Payment"


def encode_requirements_header(requirements: PaymentRequirements) -> str:
    """Encode payment requirements to header value.
    
    Uses native Rust implementation when available.
    
    Args:
        requirements: Payment requirements to encode
        
    Returns:
        Base64-encoded JSON string
    """
    if _USE_NATIVE:
        # Use Rust implementation
        native_req = NativePaymentRequirements(
            amount=requirements.amount,
            recipient=requirements.recipient,
            network=requirements.network.value if hasattr(requirements.network, 'value') else str(requirements.network),
            resource=requirements.resource,
            token=requirements.token,
            description=requirements.description,
            expires_at=requirements.expires_at,
        )
        return _native_encode_requirements(native_req)
    
    # Fallback: pure Python
    json_str = requirements.model_dump_json()
    return base64.b64encode(json_str.encode()).decode()


def decode_requirements_header(header: str) -> PaymentRequirements:
    """Decode payment requirements from header value.
    
    Uses native Rust implementation when available.
    
    Args:
        header: Base64-encoded header value
        
    Returns:
        Decoded PaymentRequirements
        
    Raises:
        ValueError: If header is invalid
    """
    if _USE_NATIVE:
        # Use Rust implementation
        native_req = _native_decode_requirements(header)
        return PaymentRequirements(
            amount=native_req.amount,
            recipient=native_req.recipient,
            network=native_req.network,
            token=None,  # TODO: handle token
            description=native_req.description,
            expires_at=native_req.expires_at,
            resource=native_req.resource,
        )
    
    # Fallback: pure Python
    try:
        json_bytes = base64.b64decode(header)
        data = json.loads(json_bytes)
        return PaymentRequirements(**data)
    except Exception as e:
        raise ValueError(f"Invalid X-Payment-Requirements header: {e}")


def encode_payment_header(payment: SignedPayment) -> str:
    """Encode signed payment to header value.
    
    Uses native Rust implementation when available.
    
    Args:
        payment: Signed payment to encode
        
    Returns:
        Base64-encoded JSON string
    """
    if _USE_NATIVE:
        # Use Rust implementation
        native_payload = NativePaymentPayload(
            amount=payment.payment.amount,
            recipient=payment.payment.recipient,
            payer=payment.payment.payer,
            chain_id=payment.payment.chain_id,
            resource=payment.payment.resource,
            nonce=payment.payment.nonce,
            expires_at=payment.payment.expires_at,
            token=payment.payment.token,
        )
        return _native_encode_payment(native_payload, payment.signature)
    
    # Fallback: pure Python
    data = {
        "payment": payment.payment.model_dump(),
        "signature": payment.signature.hex(),
    }
    json_str = json.dumps(data)
    return base64.b64encode(json_str.encode()).decode()


def decode_payment_header(header: str) -> SignedPayment:
    """Decode signed payment from header value.
    
    Uses native Rust implementation when available.
    
    Args:
        header: Base64-encoded header value
        
    Returns:
        Decoded SignedPayment
        
    Raises:
        ValueError: If header is invalid
    """
    if _USE_NATIVE:
        # Use Rust implementation
        native_payload, signature = _native_decode_payment(header)
        payload = PaymentPayload(
            amount=native_payload.amount,
            recipient=native_payload.recipient,
            payer=native_payload.payer,
            chain_id=native_payload.chain_id,
            token=None,  # TODO: handle token
            resource=native_payload.resource,
            nonce=native_payload.nonce,
            expires_at=native_payload.expires_at,
        )
        return SignedPayment(payment=payload, signature=bytes(signature))
    
    # Fallback: pure Python
    try:
        json_bytes = base64.b64decode(header)
        data = json.loads(json_bytes)
        
        payment = PaymentPayload(**data["payment"])
        signature = bytes.fromhex(data["signature"])
        
        return SignedPayment(payment=payment, signature=signature)
    except Exception as e:
        raise ValueError(f"Invalid X-Payment header: {e}")

File: core/src/protocol.rs
//! x402 protocol header encoding/decoding

use crate::{PaymentRequirements, SignedPayment, X402Error, Result};
use base64::{Engine as _, engine::general_purpose::STANDARD as BASE64};

/// Header name for payment requirements (server → client)
pub const X402_REQUIREMENTS_HEADER: &str = "X-Payment-Requirements";

/// Header name for signed payment (client → server)
pub const X402_PAYMENT_HEADER: &str = "X-Payment";

/// Encode payment requirements to header value
/// 
/// # Example
/// ```
/// use x402_core::{PaymentRequirements, encode_requirements_header, Network};
/// use alloy_primitives::{Address, U256};
/// 
/// let requirements = PaymentRequirements {
///     amount: U256::from(1000000),
///     recipient: Address::ZERO,
///     network: Network::Base,
///     token: None,
///     description: Some("API access".to_string()),
///     expires_at: None,
///     resource: "/api/data".to_string(),
/// };
/// 
/// let header = encode_requirements_header(&requirements).unwrap();
/// ```
pub fn encode_requirements_header(requirements: &PaymentRequirements) -> Result<String> {
    let json = serde_json::to_string(requirements)
        .map_err(|e| X402Error::EncodingError(e.to_string()))?;
    Ok(BASE64.encode(json.as_bytes()))
}

/// Decode payment requirements from header value
pub fn decode_requirements_header(header: &str) -> Result<PaymentRequirements> {
    let bytes = BASE64.decode(header)
        .map_err(|e| X402Error::InvalidHeader(format!("base64 decode failed: {}", e)))?;
    
    let json = String::from_utf8(bytes)
        .map_err(|e| X402Error::InvalidHeader(format!("invalid UTF-8: {}", e)))?;
    
    serde_json::from_str(&json)
        .map_err(|e| X402Error::InvalidHeader(format!("JSON parse failed: {}", e)))
}

/// Encode signed payment to header value
pub fn encode_payment_header(payment: &SignedPayment) -> Result<String> {
    let json = serde_json::to_string(payment)
        .map_err(|e| X402Error::EncodingError(e.to_string()))?;
    Ok(BASE64.encode(json.as_bytes()))
}

/// Decode signed payment from header value
pub fn decode_payment_header(header: &str) -> Result<SignedPayment> {
    let bytes = BASE64.decode(header)
        .map_err(|e| X402Error::InvalidHeader(format!("base64 decode failed: {}", e)))?;
    
    let json = String::from_utf8(bytes)
        .map_err(|e| X402Error::InvalidHeader(format!("invalid UTF-8: {}", e)))?;
    
    serde_json::from_str(&json)
        .map_err(|e| X402Error::InvalidHeader(format!("JSON parse failed: {}", e)))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Network;
    use alloy_primitives::{Address, U256};

    #[test]
    fn test_requirements_roundtrip() {
        let requirements = PaymentRequirements {
            amount: U256::from(1000000),
            recipient: Address::ZERO,
            network: Network::Base,
            token: None,
            description: Some("Test payment".to_string()),
            expires_at: Some(1700000000),
            resource: "/api/test".to_string(),
        };

        let encoded = encode_requirements_header(&requirements).unwrap();
        let decoded = decode_requirements_header(&encoded).unwrap();

        assert_eq!(decoded.amount, requirements.amount);
        assert_eq!(decoded.recipient, requirements.recipient);
        assert_eq!(decoded.resource, requirements.resource);
    }
}
